components/detournavigator/navigator.cpp

```diff
@@ -0,0 +1,1310 @@
+                                                                 solid, compact));
+                OPENMW_CHECK_DT_RESULT(rcErodeWalkableArea(&context, config.walkableRadius, compact));
+                OPENMW_CHECK_DT_RESULT(rcBuildDistanceField(&context, compact));
+                OPENMW_CHECK_DT_RESULT(rcBuildRegions(&context, compact, config.borderSize, config.minRegionArea,
+                                                      config.mergeRegionArea));
+
+                rcContourSet contourSet;
+                initContourSet(contourSet);
+                const ContourSetStackPtr contourSetPtr(&contourSet);
+                OPENMW_CHECK_DT_RESULT(rcBuildContours(&context, compact, config.maxSimplificationError, config.maxEdgeLen,
+                                                       contourSet));
+
+                if (contourSet.nconts == 0)
+                    return NavMeshData();
+
+                OPENMW_CHECK_DT_RESULT(rcBuildPolyMesh(&context, contourSet, config.maxVertsPerPoly, polyMesh));
+                OPENMW_CHECK_DT_RESULT(rcBuildPolyMeshDetail(&context, polyMesh, compact, config.detailSampleDist,
+                                                             config.detailSampleMaxError, polyMeshDetail));
+            }
+
+            for (int i = 0; i < polyMesh.npolys; ++i)
+                if (polyMesh.areas[i] == RC_WALKABLE_AREA)
+                    polyMesh.flags[i] = 1;
+
+            dtNavMeshCreateParams params;
+            params.verts = polyMesh.verts;
+            params.vertCount = polyMesh.nverts;
+            params.polys = polyMesh.polys;
+            params.polyAreas = polyMesh.areas;
+            params.polyFlags = polyMesh.flags;
+            params.polyCount = polyMesh.npolys;
+            params.nvp = polyMesh.nvp;
+            params.detailMeshes = polyMeshDetail.meshes;
+            params.detailVerts = polyMeshDetail.verts;
+            params.detailVertsCount = polyMeshDetail.nverts;
+            params.detailTris = polyMeshDetail.tris;
+            params.detailTriCount = polyMeshDetail.ntris;
+            params.offMeshConVerts = nullptr;
+            params.offMeshConRad = nullptr;
+            params.offMeshConDir = nullptr;
+            params.offMeshConAreas = nullptr;
+            params.offMeshConFlags = nullptr;
+            params.offMeshConUserID = nullptr;
+            params.offMeshConCount = 0;
+            params.walkableHeight = getHeight(agentHalfExtents);
+            params.walkableRadius = getRadius(agentHalfExtents);
+            params.walkableClimb = getMaxClimb();
+            rcVcopy(params.bmin, polyMesh.bmin);
+            rcVcopy(params.bmax, polyMesh.bmax);
+            params.cs = config.cs;
+            params.ch = config.ch;
+            params.buildBvTree = true;
+            params.userId = 0;
+            params.tileX = tileX;
+            params.tileY = tileY;
+            params.tileLayer = 0;
+            rcVcopy(params.bmin, config.bmin);
```

Comment:
@elsid I think I actually managed to fix the tiled navmesh, went through it line by line comparing it to the code in the sample, wrapping my head around it as I went, and found something I think you missed when you copied and pasted the code from the code to generate a single tile. This line and the line after it need to be removed, this is basically resetting the bounds to be using the bounds set at the beginning of the function, which is where you changed the bounds, so things became all deformed because the bounds were wrong. Running this with the lines removed made the nav mesh be correct: 
![fixed-navmesh](https://user-images.githubusercontent.com/791804/37809865-7f01be04-2e0e-11e8-80b1-16fabc264a6a.png)

Reply:
Thanks, I've spent a lot of time to check everything above, but missed this one.

I will take a look to TileCache.

Reply:
The Sample_TempObstacles.cpp in the RecastDemo is the demo implementation for TileCache, supposedly it is more efficient to create navmeshes this way, as you incur a one time rasterization step then can use that intermediate part to create the navmesh. Still seems to take as much time in the demo as it has to do the rasterization but the time to generate say 9 exterior cells it listed at 245ms or something, so it might be a good idea to look into. There are also things within it to do compression of the data using FastLZ4 to pretty massively cut down the size of the data in memory. It supposedly will take a bit more memory on startup and such with the cache, but should improve the generation of the navmesh. Hopefully it will work out for our needs, I've taken a look at some of the code and it seemed a bit different but still not 100% familiar with all of the recast stuff, and my C/C++ is incredibly rusty so sorry I can't be more of help.